### PR TITLE
feat: Write Chunks to Parquet Files in Object Store

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #

This is initial implementation of [#1801](https://github.com/influxdata/influxdb_iox/issues/1081). The two major tasks in here is to implement:
1. Function` load_chunk_to_object_store` that write closed mutable buffer chunks to object store in parquet format, one file per table of a chunk. The file name will be: `<writer id>/<database>/data/<partition key>/<chunk id>/<tablename>.parquet`.
2. Implement parquet_file crate that plays similar role as mutable buffer and read buffer but for parquet files.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Comment by NGA-TRAN** [comment]
@alamb, @tustvold , @mkmik , and @pauldix This is not completed but I have put all todos in places I think I need to cover. Let me know what you think. Also, there will be a lot more code so if there are no major concerns, I want to have this framework merged to avoid long code review in one PR.

@alamb If you are available tomorrow/Thursday morning, I would like to pick our brain on how to store partitions of parquet files in the newly created Catalog.

File: parquet_file/src/parquet_chunk.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct MemWriter {
+    mem: Arc<Mutex<Cursor<Vec<u8>>>>,
+}
+
```
Comment (NGA-TRAN) [inline parquet_file/src/parquet_chunk.rs:101]: The code from here down is moved from snapshot.rs

File: server/src/db.rs [hunk 2]
```diff
@@ -372,6 +407,108 @@ impl Db {
         Ok(DBChunk::snapshot(&chunk))
     }
 
+    pub fn load_chunk_to_object_store(
```
Comment (NGA-TRAN) [inline server/src/db.rs:410]: This is the function that start writing to object store

File: data_types/src/database_rules.rs [hunk 3]
```diff
@@ -76,6 +76,10 @@ impl DatabaseRules {
     pub fn new() -> Self {
         Self::default()
     }
+
+    pub fn db_name(&self) -> String {
```
Comment (tustvold) [inline data_types/src/database_rules.rs:80]: Why is this necessary when the name member is public?

File: parquet_file/src/parquet_chunk.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    ) -> Result<()> {
+        // Full path of the file in object store
+        //    <writer id>/<database>/data/<partition key>/<chunk id>/<table
+        // name>.parquet
+
+        // TODO: write a function to return an object_store::path::Path's whose
+        // root value is already "<writer id>/"
+        // This function must be somewhere in server so it has knowledge of the
+        // writer_id
+        let store = Arc::new(ObjectStore::new_in_memory(InMemory::new()));
```
Comment (tustvold) [inline parquet_file/src/parquet_chunk.rs:58]: I would have expected a reference to an object store to be passed as an argument?

File: parquet_file/src/parquet_chunk.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        let mut path = store.new_path();
+        path.push_dir("100");
+
+        // Continue building the path
```
Comment (tustvold) [inline parquet_file/src/parquet_chunk.rs:62]: Minor niggle - this should probably be a function

File: parquet_file/src/parquet_chunk.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            id: chunk_id,
+            object_store_paths: vec![],
+        }
+    }
+
+    pub async fn write_to_object_store(
+        &self,
+        db_name: String,
+        table_name: String,
+        cursor: MemWriter,
```
Comment (tustvold) [inline parquet_file/src/parquet_chunk.rs:48]: It can always be moved later, but I'd have perhaps expected the logic and mechanics of converting data to Parquet to live in parquet_file as opposed to in Db and this then passed what is, if my understanding is correct, an in-memory buffer with a serialized parquet file.

I'd perhaps expect a non-member function that would take a read buffer chunk and an object store, and return a ParquetChunk or something?

File: server/src/db/chunk.rs [hunk 7]
```diff
@@ -68,7 +69,9 @@ pub enum DBChunk {
         partition_key: Arc<String>,
         chunk_id: u32,
     },
-    ParquetFile, // TODO add appropriate type here
+    ParquetFile {
```
Comment (tustvold) [inline server/src/db/chunk.rs:72]: Do we expect ParquetFiles do be directly queryable? I had, perhaps incorrectly, assumed that at least initially we'd hydrate the ParquetFile into the read buffer and query that?

File: parquet_file/src/parquet_chunk.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+     * the table meta data can be extracted from the parquet file so I do not
+     * store them for now but we might need them here for easy-to-use reason. Will see */
+}
+
+impl ParquetChunk {
+    pub fn new(part_key: String, chunk_id: u32) -> Self {
+        Self {
+            partition_key: part_key,
+            id: chunk_id,
+            object_store_paths: vec![],
```
Comment (e-dard) [inline parquet_file/src/parquet_chunk.rs:40]: Perhaps this doesn't matter right now but one nice thing you can do in Rust is something like:

```rust
pub fn new(part_key: String, chunk_id: u32) -> Self {
        Self {
            partition_key: part_key,
            id: chunk_id,
            Self::default()...
        }
```

To do this you will need to add a derive `Default`on `ParquetChunk`. The advantage of this is if you add new fields to `ParquetChunk` in the future and you don't need to set them on `new` their default values will be set for you without you refactoring `new`

File: server/src/db/chunk.rs (hunk 7 above)
Comment (e-dard) [reply server/src/db/chunk.rs:72]: Ideally I think we may want to execute against parquet files. So really I think we want to do both. It only probably makes sense to pay the price to move to Read Buffer if that data is going to be queried multiple times and I suppose that's not always the case.

Further, if someone wants to query across TB of parquet data I think we will need a better story than move all of that through the Read buffer.

File: parquet_file/src/parquet_chunk.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Error writing to object store: {}", source))]
+    WritingToObjectStore { source: object_store::Error },
+}
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Clone)]
+pub struct ParquetChunk {
```
Comment (e-dard) [inline parquet_file/src/parquet_chunk.rs:21]: What do you think about calling this `Chunk`? The fact it's a parquet backed chunk can be derived from the module it's in (the parquet module).

File: parquet_file/src/parquet_chunk.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        store
+            .put(
+                &path,
+                futures::stream::once(async move { stream_data }),
+                Some(len),
+            )
+            .await
+            .context(WritingToObjectStore)?;
+
+        // TODO: Add this path into this chunk object_store_paths list
```
Comment (e-dard) [inline parquet_file/src/parquet_chunk.rs:89]: Is this something you need help with?

**Review by e-dard — COMMENTED** [review]
This is looking like a good direction @NGA-TRAN. I had a few comments but generally this makes sense to me 🎉

File: server/src/db/chunk.rs (hunk 7 above)
Comment (alamb) [reply server/src/db/chunk.rs:72]: Yes I agree that being able to query directly from parquet files (cached locally) is important for he reasons @e-dard  says:

1. Query while data is being loaded  into the read buffer (have to load all columns, encode them, etc) 
2. Datasets where we don't have the memory to be able to keep everything in the read buffer at once

File: data_types/src/database_rules.rs (hunk 3 above)
Comment (alamb) [reply data_types/src/database_rules.rs:80]: An alternate that doesn't require a copy might look like

```
    pub fn db_name(&self) -> &str {
      &self.name
    }
```

if we want to move towards using accessor functions rather than public fields

File: server/src/db.rs [hunk 11]
```diff
@@ -372,6 +407,108 @@ impl Db {
...
+                chunk_id,
+            })?
+        };
+
+        // update the catalog to say we are processing this chunk and
+        // then drop the lock while we do the work
+        let mb_chunk = {
+            let mut chunk = chunk.write();
+
+            // TODO: Make sure this is set to the corresponding "moving to object store
```
Comment (alamb) [inline server/src/db.rs:437]: 👍  makes sense to me

File: server/src/db.rs [hunk 12]
```diff
@@ -372,6 +407,108 @@ impl Db {
...
+                continue;
+            }
+
+            // For now batches.len() is always 1 when reach here but in the future if is it
+            // >1, we need to verify whether schema of all RecordBatch of the
+            // same table of this chunk has the same schema
+            let schema = batches[0].schema();
+
+            // memory for the parquet content of this table
+            let cursor = MemWriter::default();
```
Comment (alamb) [inline server/src/db.rs:476]: I can't remember why we needed to write this to memory rather than streaming it through, but I think it is good enough for now and we can reduce the peak memory usage subsequently

File: server/src/db.rs [hunk 13]
```diff
@@ -372,6 +407,108 @@ impl Db {
...
+            writer.close().context(ClosingParquetWriter)?;
+
+            // TODO: put this function under a tokio
+            // Put the file to object store
+            // parquet_chunk.write_to_object_store(self.db_name(), stats.name,
+            // cursor);
+        }
+
+        // TOOD: add the parquet_chunk into a member of this Db (which maybe Catalog or
+        // so) Need to talk with Andrew and Raphael about this
```
Comment (alamb) [inline server/src/db.rs:492]: Yeah, I recommend adding a new state to `catalog::chunk::ChunkState` called `ObjectStore` or something which holds a `ParquetChunk`

File: parquet_file/src/parquet_chunk.rs (hunk 4 above)
Comment (alamb) [reply parquet_file/src/parquet_chunk.rs:58]: I thought there was already an `ObjectStore` reference on `Db` but it appears that is not the case: https://github.com/influxdata/influxdb_iox/blob/b0e21e5f9e8083b6fbf7a46590e8192a26d35daa/server/src/db.rs#L160

However, it lives on the `Server` object: https://github.com/influxdata/influxdb_iox/blob/b0e21e5f9e8083b6fbf7a46590e8192a26d35daa/server/src/lib.rs#L193 so we just need to pass it into the Db creation, probably

File: parquet_file/src/parquet_chunk.rs (hunk 6 above)
Comment (alamb) [reply parquet_file/src/parquet_chunk.rs:48]: i think it makes sense to get a `RecordBatchStream` here and handle writing to parquet in `parquet_file` -- I would personally suggest not putting dependencies on the mutable buffer in here. I think this is consistent with what @tustvold  is suggesting as well.

**Review by alamb — COMMENTED** [review]
I think this is a great step forward! I think the only major thing missing are some new states   in `catalog::chunk::ChunkState` that hold a `ParquetFile`  -- perhaps we can add that, write some tests, and then merge this PR and iterate on additional comments afterwards.

Nice work @NGA-TRAN !

**Review by alamb — APPROVED** [review]

File: parquet_file/src/parquet_chunk.rs (hunk 6 above)
Comment (tustvold) [reply parquet_file/src/parquet_chunk.rs:48]: @alamb that would be even better - yes 🙂

File: parquet_file/src/parquet_chunk.rs (hunk 9 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:21]: Yes, I was debating and I think `Chunk` is a more consistent name

File: parquet_file/src/parquet_chunk.rs (hunk 8 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:40]: Great idea

File: parquet_file/src/parquet_chunk.rs (hunk 6 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:48]: Sounds good. Actually this is what I wanted to do at first be then I see what we did with read buffer and followed. Will change it.

File: parquet_file/src/parquet_chunk.rs (hunk 4 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:58]: Yes, I was looking for it. I took a todo note in the Parquet_file Crete to do this

File: parquet_file/src/parquet_chunk.rs (hunk 5 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:62]: Yes, all the object store's path building will be in its own function, including the root writer id

File: parquet_file/src/parquet_chunk.rs (hunk 10 above)
Comment (NGA-TRAN) [reply parquet_file/src/parquet_chunk.rs:89]: I think if I know where we manage parquet partitionss, I might be able to figure this out or will ask for help at that time
